pyteal/ast/abi/address.py

```diff
@@ -1,33 +1,120 @@
     def __init__(self) -> None:
-        super().__init__(ByteTypeSpec(), ADDRESS_LENGTH)
+        super().__init__(ByteTypeSpec(), AddressLength.Bytes.value)
 
     def new_instance(self) -> "Address":
         return Address()
 
     def __str__(self) -> str:
         return "address"
 
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, AddressTypeSpec)
+
 
 AddressTypeSpec.__module__ = "pyteal"
 
 
-class Address(StaticArray):
+class Address(StaticArray[Byte, Literal[32]]):
     def __init__(self) -> None:
-        super().__init__(AddressTypeSpec(), ADDRESS_LENGTH)
+        super().__init__(AddressTypeSpec())
 
     def type_spec(self) -> AddressTypeSpec:
         return AddressTypeSpec()
 
     def get(self) -> Expr:
         return self.stored_value.load()
 
+    def set(
+        self,
+        value: Union[
+            Sequence[Byte],
+            StaticArray[Byte, Literal[32]],
+            ComputedValue[StaticArray[Byte, Literal[32]]],
+            "Address",
+            str,
+            bytes,
+            Expr,
+        ],
+    ):
+
+        match value:
+            case ComputedValue():
+                pts = value.produced_type_spec()
+                if pts == AddressTypeSpec() or pts == StaticArrayTypeSpec(
+                    ByteTypeSpec(), AddressLength.Bytes.value
+                ):
+                    return value.store_into(self)
+
+                raise TealInputError(
+                    f"Got ComputedValue with type spec {pts}, expected AddressTypeSpec or StaticArray[Byte, Literal[32]]"
+                )
+            case BaseType():
+                if (
+                    value.type_spec() == AddressTypeSpec()
+                    or value.type_spec()
+                    == StaticArrayTypeSpec(ByteTypeSpec(), AddressLength.Bytes.value)
+                ):
+                    return self.stored_value.store(value.stored_value.load())
+
+                raise TealInputError(
+                    f"Got {value} with type spec {value.type_spec()}, expected AddressTypeSpec"
+                )
+            case str():
+                if len(value) == AddressLength.String.value:
+                    return self.stored_value.store(Addr(value))
+                raise TealInputError(
+                    f"Got string with length {len(value)}, expected {AddressLength.String.value}"
+                )
+            case bytes():
+                if len(value) == AddressLength.Bytes.value:
+                    return self.stored_value.store(Bytes(value))
+                raise TealInputError(
+                    f"Got bytes with length {len(value)}, expected {AddressLength.Bytes.value}"
+                )
+            case Expr():
+                return self.stored_value.store(value)
+            case CollectionSequence():
+                # TODO: mypy thinks its possible for the type of `value` here to be
+                # Union[Sequence[Byte], str, bytes] even though we check above?
```

Comment:
This confuses me, is it an issue with mypy or am I just confused about how this match case should work?

Reply:
@barnjamin Take my thoughts with a grain of salt - I think my understanding of the Python type system is still limited.

My guess:   I think it's a language limitation rather than a user error.

Possible workaround: 
* Capture the matched pattern as shown in https://peps.python.org/pep-0636/#capturing-matched-sub-patterns.
* I think the proposal is succinct and _seems_ to avoid introducing additional classes of error.  
* Though I'm hesitant to use the approach more broadly because the IDE (at least Pycharm) resolves `cs` as type `Any`.  Presumably the lack of type inference will be improved over time.  For the mentioned issue, I'd be comfortable applying it if you feel it's an improvement.

```
            case CollectionSequence() as cs:
                return super().set(cs)
```

Reply:
Mypy is wrong here. Eventually they should get this right, but in the meantime you should explicitly cast the variable to the right type (i.e. `cast(Sequence[Byte], value)`).

FWIW VS Code (PyLance) correctly types `value` as `Sequence[Byte]` for me.